packages/starknet-snap/src/utils/superstruct.ts

```diff
@@ -91,3 +94,101 @@ export const CairoVersionStruct = enums([CAIRO_VERSION, CAIRO_VERSION_LEGACY]);
+ *
+ * @param predefinedProperties - A Superstruct schema defining the base set of properties that are expected.
+ * @param additionalPropertyTypes - A Superstruct schema that defines the types for any additional properties.
+ * Defaults to `any`, allowing any additional properties.
+ * @returns A dynamic struct that first validates against the predefined properties and then
+ * includes any additional properties that match the `additionalPropertyTypes` schema.
+ */
+export const createStructWithAdditionalProperties = (
+  predefinedProperties: Struct<any, any>,
+  additionalPropertyTypes: Struct<any, any> = any(),
+) => {
+  return dynamic((value) => {
+    if (typeof value !== 'object' || value === null) {
+      return predefinedProperties;
+    }
+
+    const additionalProperties = Object.keys(value).reduce<
+      Record<string, Struct>
+    >((schema, key) => {
+      if (!(key in predefinedProperties.schema)) {
+        schema[key] = additionalPropertyTypes;
+      }
+      return schema;
+    }, {});
+
+    return assign(predefinedProperties, object(additionalProperties));
+  });
+};
+
+// Define the types you expect for additional properties
+const additionalPropertyTypes = union([string(), number(), any()]);
+
+const DeclarePayloadStruct = createStructWithAdditionalProperties(
+  object({
+    contract: union([any(), string()]),
+    classHash: optional(string()),
+    casm: optional(any()),
+    compiledClassHash: optional(string()),
+  }),
+  additionalPropertyTypes,
+);
+
+const InvokePayloadStruct = createStructWithAdditionalProperties(
+  object({
+    contractAddress: string(),
+    calldata: optional(any()), // Assuming RawArgs or Calldata can be represented as any or string
+    entrypoint: optional(string()), // Making entrypoint optional as it was mentioned in the example
+  }),
+  additionalPropertyTypes,
+);
+
+const DeclareTransactionStruct = object({
+  type: enums([TransactionType.DECLARE]),
+  payload: optional(DeclarePayloadStruct),
+});
+
+const DeployTransactionStruct = object({
+  type: enums([TransactionType.DEPLOY]),
+  payload: optional(any()),
+});
+
+const DeployAccountTransactionStruct = object({
+  type: enums([TransactionType.DEPLOY_ACCOUNT]),
+  payload: optional(any()),
+});
+
+const InvokeTransactionStruct = object({
+  type: enums([TransactionType.INVOKE]),
+  payload: optional(InvokePayloadStruct),
+});
+
+export const InvocationStruct = union([
+  DeclareTransactionStruct,
+  DeployTransactionStruct,
+  DeployAccountTransactionStruct,
+  InvokeTransactionStruct,
+]);
+
+export const UniversalDetailsStruct = object({
+  nonce: optional(any()),
+  blockIdentifier: optional(any()),
+  maxFee: optional(any()),
+  tip: optional(any()),
+  paymasterData: optional(array(any())),
+  accountDeploymentData: optional(array(any())),
+  nonceDataAvailabilityMode: optional(any()),
+  feeDataAvailabilityMode: optional(any()),
+  version: optional(enums(['0x2', '0x3'])),
```

Comment:
```suggestion
  version: optional(TxVersionStruct)),
```

Reply:
actually i have already declare  this in https://github.com/Consensys/starknet-snap/pull/333/files
and im not using any too, preherps we can merge them together